yoyodyne/models/pointer_generator.py

```diff
@@ -444,3 +447,238 @@ def _reshape_hiddens(
+        target_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        """Runs the decoder in one step.
+
+        This will work on any sequence length, and returns output probabilities
+        for all targets, meaning we can use this method for greedy decoding,
+        wherein only a single new token is decoded at a time, or for teacher-forced
+        training, wherein all tokens can be decoded in parallel with a diagonal mask.
+
+        Args:
+            encoder_outputs (torch.Tensor): Encoded output representations.
+            source_mask (torch.Tensor): Mask for the encoded source tokens.
+            source_indices (torch.Tensor): Source token vocabulary ids.
+            target_tensor (torch.Tensor): Target token vocabulary ids.
+            target_mask (torch.Tensor): Mask for the target tokens.
+
+        Returns:
+            torch.Tensor: Output probabilities of the shape 
+                B x target_seq_len x target_vocab_size, where
+                target_seq_len is inferred form the target_tensor.
+        """
+        decoder_output = self.decoder(
+            encoder_outputs, source_mask, target_tensor, target_mask
+        )
+        tgt_embeddings = decoder_output.embeddings
+        decoder_output = decoder_output.output
+        # Outputs from multi-headed attention from each decoder step to
+        # the encoded inputs.
+        # Values have been averaged over each attention head.
+        # -> B x tgt_seq_len x src_seq_len
+        mha_outputs = self.decoder.attention_output.outputs[0]
+        # Clears the stored attention result.
+        self.decoder.attention_output.clear()
+        logits = self.classifier(decoder_output)
+        output_probs = nn.functional.softmax(logits, dim=2)
+        # -> B x target_seq_len x target_vocab_size.
+        ptr_probs = torch.zeros(
+            mha_outputs.size(0),
+            mha_outputs.size(1),
+            self.target_vocab_size,
+            device=self.device,
+            dtype=mha_outputs.dtype,
+        )
+        # Repeats the source indices for each target.
+        # -> B x tgt_seq_len x src_seq_len
+        repeated_source_indices = source_indices.unsqueeze(1).repeat(
+            1, mha_outputs.size(1), 1
+        )
+        # Scatters the attention weights onto the ptr_probs tensor
+        # at their vocab indices in order to get outputs
+        # that match the indexing of the generation probability.
+        ptr_probs.scatter_add_(
+            2, repeated_source_indices, mha_outputs
+        )
+        # A matrix of 'context' vectors from applying attention 
+        # to the encoder representations wrt each decoder step.
+        context = torch.bmm(mha_outputs, encoder_outputs)
+        # Probability of generating (from output_probs).
+        gen_probs = self.generation_probability(
+            context, decoder_output, tgt_embeddings
+        )
+        ptr_scores = (1 - gen_probs) * ptr_probs
+        gen_scores = gen_probs * output_probs
```

Comment:
consider using `.mul_` here

Reply:
Is it faster?

Reply:
Or are you suggesting because it is in-place?

Reply:
Yeah, I am suggesting in-place because it shouddl be faster (clearer too maybe?). the way you wrote it:

* creates a new tensor with (1 - gen_probs) * ptr_probs in it
* assigns it to `ptr_scores`, which triggers atomic increments/decremetns in the ref counter. since the ref count for the older data that `ptr_probs` pointed to should now be zero, it will be marked for (but not yet undergo) GC on the next cycle
* ultimately the GC will also delete the old ptr_probs data

Only the tensor multiplication is necessary if you use the in-place one.

Reply:
Ok. I just need to find the op for a vector * scalar. I get an error with `_mul` because they are not the same shape.

Reply:
I am wrong, I just was calling mul_ on the vector first.

Reply:
These in-place operations seem to break backprop, though I do not understand the error message.

Reply:
I assume its b/c we are overwriting something on those variables that needs to be stored in the CG for backprop, and no longer is?

Reply:
FWIW we have used these elsewhere in similar contexts, I believe, and no related bugs yet. If this really does break it I wonder if we're doing "silly scoping" somewhere---i.e., something isn't being passed in or something like that.

Reply:
> something isn't being passed in or something like that.

Can you explain more here? I can look into this a bit harder if it is a symptom of another issue.

Reply:
For instance maybe we wrote something that is conceptually similar to:

```python
def foo(bar):
   return baz + bar   # where is baz from? 
```

Barring that I can't imagine how this would influence the computation graph.

Reply:
You can always just leave this as a `# TODO: consider in-place operations; using .mul_ here seems to cause issues.` or something like that.

Reply:
Ok, I will do that for now.

Reply:
pytorch is annoyingly inconsistent with in place operation and maintaining position in the computation graph (also some weird stuff regarding operation managed by ptl AMP for 16 vs 32 bit precision). Safer is usually to just cast to a new tensor (pytorch is usually reliable with only duplicating data when needed).